### PR TITLE
Implement library for custom scaffolder actions.

### PR DESCRIPTION
**Pull request by Xantier**
* Add functionality to receive custom scaffolder action triggering
* Add context provider to handle action running
* Add workspace manager to retrieve and upload wanted scaffolder workspace from/to S3 based on provided presigned URLs
* Add CustomScaffolderAction class to wrap the needed endpoint calls to create a simpler library interface
* Add docs

### CODE REVIEW

File: src/lib/RoadieAgentReceiver.ts [hunk 1]
```diff
@@ -76,4 +83,38 @@ export class RoadieAgentReceiver {
       );
     });
   }
+
+  private registerScaffolderAction(
+    configuration: ScaffolderActionAgentConfiguration,
+    app: Express,
+  ) {
+    // Specifying routes explicitly
+    app.get(`/scaffolder-action/${configuration.name}`, (req, res) => {
```
Comment (Xantier) [inline src/lib/RoadieAgentReceiver.ts:92]: Entrypoint to the Agent functionality. Triggers the action by creating a context and helper wrapper for it, within `CustomScaffolderAction`

File: src/lib/scaffolderAction/CustomScaffolderAction.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      brokerClientUrl,
+      payload: { localWorkspacePath, body: payload.body },
+    });
+  }
+
+  async start() {
+
+    this.logger.info('Starting a custom scaffolder action in 5 seconds:');
+    // TODO: Figure out read-after write consistency
+    await new Promise(resolve => setTimeout(resolve, 5000));
```
Comment (Xantier) [inline src/lib/scaffolderAction/CustomScaffolderAction.ts:45]: S3 is strongly consistent, no need for this anymore.

File: src/lib/scaffolderAction/CustomScaffolderAction.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    try {
+      // TODO: filesystem handling: Create folder and temp workspace
+      console.log(this.workspaceUrl);
+      if (this.workspaceUrl && this.workspaceUrl !== 'undefined') {
+        await downloadFile(this.workspaceUrl, this.localWorkspacePath);
```
Comment (Xantier) [inline src/lib/scaffolderAction/CustomScaffolderAction.ts:50]: Retrieves the stored workspace using a presigned URL

File: src/lib/scaffolderAction/CustomScaffolderAction.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      }
+
+      await this.configuration.handler(this.context);
+
+      if (this.workspaceUrl && this.workspaceUrl !== 'undefined') {
+        const workspaceUrl = (await generateAndStreamZipfileToS3(this.localWorkspacePath, this.actionId))!;
```
Comment (Xantier) [inline src/lib/scaffolderAction/CustomScaffolderAction.ts:56]: Stores the modified workspace into S3. (Needs a presigned URL to be passed into it one the payload still)

File: src/lib/scaffolderAction/CustomScaffolderActionContext.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  }) {
+    this.brokerClientUrl = brokerClientUrl;
+    this.actionId = actionId;
+    this.payload = payload;
+    this.workspacePath = payload.localWorkspacePath;
+    this.logger = getLogger('RoadieAgentForwarder');
+  }
+
+
+  async log(content: string, context?: Record<string, string>) {
```
Comment (Xantier) [inline src/lib/scaffolderAction/CustomScaffolderActionContext.ts:29]: Simple custom scaffolder context which exposes a logger, workspace and the payload passed into when the action was triggered

File: src/lib/entityProvider/constants.ts
Comment (Xantier) [inline src/lib/entityProvider/constants.ts:1]: entityProvider side of the agent is untouched, just moved to a folder

File: src/lib/scaffolderAction/workspaceHandler.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  try {
+    await zipDirectory(workspacePath, `${workspacePath}.zip`);
+    logger.info('Uploading zip to S3');
+    const response = await uploadWorkspace(
+      putPresignUrl,
+      `${workspacePath}.zip`,
+    );
+
+    logger.info('File uploaded');
+    const etag = (response?.headers?.etag as string)?.replaceAll('"', '');
```
Comment (Xantier) [inline src/lib/scaffolderAction/workspaceHandler.ts:105]: AWS S3 produces etags that are wrapped in `"`. It's implemented according to spec and will never change

**Review by roadgork — APPROVED** [review]
can't see any problems

**Pull request merged**